qiskit/circuit/singleton.py

```diff
@@ -258,37 +378,118 @@ def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
         return cls
 
     def __call__(cls, *args, _force_mutable=False, **kwargs):
-        if not _force_mutable and not args and not kwargs:
-            # This class attribute is created by the singleton-creation base classes'
-            # `__init_subclass__` methods; see `_impl_init_subclass`.  We can only be within this
-            # `__call__` after the class (e.g. `XGate`) was created, and `XGate._singleton_instance`
-            # is created during the Python-standard `XGate = type("XGate", bases, namespace)` step.
-            return cls._singleton_instance
+        if _force_mutable:
+            return super().__call__(*args, **kwargs)
+        if not args and not kwargs:
+            # This is a fast-path to handle constructions of the form `XGate()`, which is the
+            # idiomatic way of building gates during high-performance circuit construction.  If
+            # there are any arguments or kwargs, we delegate to the overridable method to
+            # determine the cache key to use for lookup.
+            return cls._singleton_default_instance
+        if (key := cls._singleton_lookup_key(*args, **kwargs)) is not None:
+            try:
+                singleton = cls._singleton_static_lookup.get(key)
+            except TypeError:
```

Comment:
Is the try/except here faster than the `None` default on `get()`? If I were using the try except pattern I would have just don `cls._singleton_static_lookup.get[key]` instead of calling `get()`

Reply:
This isn't about catching the `None` (which would raise `KeyError` not `TypeError`), it's about catching the case that the key returned was unhashable.  Since the key is highly likely to contain arbitrary user input, it's not fair for key implementors to manually validate that the user input is hashable, and at best they'd be duplicating work that we have to do here with the try-and-see approach anyway.

For example:
```python
>>> {}.get([], None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```

And the reason we need this, is say we're defining the key for `CXGate`, something like:
```python
def key(label=None, ctrl_state=None):
    return (ctrl_state,) if label is None else None
```
now the user does something pathological like `CXGate(ctrl_state=[1, 2, 3])`.  If we didn't `try`/`catch`, they'd get a crazy error message that really doesn't show the problem.  With this form, we gracefully build a mutable instance for them, and let `CXGate.__init__` handle the type error with its normal mechanisms.

Reply:
Oh, unless I misunderstood you - since we're already doing `try`/`except` (which in Python 3.10+ has I think zero or _very_ near zero cost), we could skip the Python-space function call and catch `KeyError` on the `None` instead?

Reply:
Yeah, I was thinking to avoid the extra call overhead of `get()`, but I agree it'd probably be ugly enough and such a small difference we wouldn't be able to measure it that's it not worth it.